### PR TITLE
Fix Consent API integration with OSA (trunk-based)

### PR DESCRIPTION
**Pull request by tomalec**
This is a follow-up from https://github.com/woocommerce/woocommerce/pull/39701
and alternative version of https://github.com/woocommerce/woocommerce/pull/41856
that is based on `trunk` and **does not** require https://github.com/woocommerce/woocommerce/pull/41690 .

It still uses a hardcoded fields list, disregarding `wc_order_attribution_tracking_fields` hook.
If this branch gets merged #41690 will need to be updated.

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

- [Stop tracking order attribution when the consent is revoked on the same page.](https://github.com/woocommerce/woocommerce/commit/60d16fe9f6a83836830c83c19352144ae147b8a4) 
    Switch `wp_consent_type_defined` to VanillaJS event, as this is what the latest Complianz dispatches.
    Remove jQuery dependency.
- [Update checkout block data when consent is given/revoked on the same page](https://github.com/woocommerce/woocommerce/pull/41856/commits/c2deff5176d9d44671ea4a09301f9ea116a33aac) 
c2deff5

Closes # .

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

0. Install WP Consent API plugin
1. Setup another plugin that uses WP Consent API, like Complianz

##### Denied by default
When you install Compilanz, the marketing data collection should be disabled by default, so even if you do not click anything in the cookie banner, `wc_order_attribution.params.allowTracking` should be set to `false` and no order attribution data should be collected. 

:lady_beetle: This integration was broken as the `WPConsentAPI` register was not called, and `-blocs` integration disregarded `wc_order_attribution.params.allowTracking` setting.

0. Make sure Complianz is set up in `opt-in` mode, for example by picking GDPR(EU) at `/wp-admin/admin.php?page=cmplz-wizard&step=1&section=1`
1. Open shop page in incognito mode. Do not click anything in the cookie banner
2. Add something to the cart and proceed to the block/classic checkout
3. Open dev tools
   - For classic checkout: Check that all `document.querySelectorAll('input[name^=wc_order_attribution_]');` inputs have empty values
   - For block checkout, monitor network tab and `/checkout` XHR
   - `wc_order_attribution.params.allowTracking` should be set to `false` 
4. "Place order"
	- For classic checkout, confirm that submitted POST **does not** contain `wc_order_attribution_*`  values
	- For block checkout, the `woocommerce/order-attribution` extension payload contains all `null`s 
6. Check that the placed order has no order attribution data in the `/wp-admin/admin.php?page=wc-orders&action=edit&id=`

##### Consent granted on the checkout page
:bug: This behavior was covered by JS code for classic checkout, so only registering `WPConsentAPI` would be all we need. But the behavior was broken for blocks

1. Start a new incognito session. Do not click anything in the cookie banner
2. Add something to the cart and proceed to the block/classic checkout
3. Open dev tools
   - For classic checkout: Check that all `document.querySelectorAll('input[name^=wc_order_attribution_]');` inputs have empty values
   - For block checkout, monitor network tab and `/checkout` XHR
   - `wc_order_attribution.params.allowTracking` should be set to `false` 
4. Grant "marketing" consent without reloading the page
   - For classic checkout: Check that all `document.querySelectorAll('input[name^=wc_order_attribution_]');` inputs have set values
   - `wc_order_attribution.params.allowTracking` should be set to `true` 
4. "Place order"
	- For classic checkout, confirm that submitted POST **does** contain `wc_order_attribution_*`  values
	- For block checkout, the `woocommerce/order-attribution` extension payload contains OSA data
6. Check that the placed order has order attribution data in the `/wp-admin/admin.php?page=wc-orders&action=edit&id=`

##### Consent granted before
1. Start a new incognito session. Do grant marketing consent in the cookie banner, reload the page
2. Add something to the cart and proceed to the block/classic checkout
3. Open dev tools
   - For classic checkout: Check that all `document.querySelectorAll('input[name^=wc_order_attribution_]');` inputs have set  values
   - For block checkout, monitor the network tab and `/checkout` XHRvalues
   - `wc_order_attribution.params.allowTracking` should be set to `true` 
4. "Place order"
	- For classic checkout, confirm that submitted POST **does** contain `wc_order_attribution_*`  values
	- For block checkout, the `woocommerce/order-attribution` extension payload contains OSA data
6. Check that the placed order has order attribution data in the `/wp-admin/admin.php?page=wc-orders&action=edit&id=`

##### Consent revoked on the checkout page
I didn't manage to simulate this behavior with the Compilanz UI, but theoretically, API supports this, so I believe we should be prepared for such behavior from another plugin.

To support this behavior, we need the majority of JS changes from this PR.

1. Start a new incognito session. Do grant marketing consent in the cookie banner, reload the page, or play with granting it at any later step before revoking it.
2. Add something to the cart and proceed to the block/classic checkout
3. Open dev tools
   - For classic checkout: Check that all `document.querySelectorAll('input[name^=wc_order_attribution_]');` inputs have set values
   - For block checkout, monitor the network tab and `/checkout` XHRvalues
   - `wc_order_attribution.params.allowTracking` should be set to `true` 
4. :yellow_circle: Simulate revoking the consent 
   ```js
   document.body.dispatchEvent(new CustomEvent('wp_listen_for_consent_change',{detail: {marketing: 'no'}}));
   ```

   - For classic checkout: Check that all `document.querySelectorAll('input[name^=wc_order_attribution_]');` inputs have empty values
   - `wc_order_attribution.params.allowTracking` should be set to `false` 
4. "Place order"
	- For classic checkout, confirm that submitted POST **does not** contain `wc_order_attribution_*`  values
	- For block checkout, the `woocommerce/order-attribution` extension payload contains all `null`s
6. Check that the placed order has no order attribution data in the `/wp-admin/admin.php?page=wc-orders&action=edit&id=`

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/Internal/Orders/OrderAttributionController.php
Comment (layoutd) [inline plugins/woocommerce/src/Internal/Orders/OrderAttributionController.php]: I think this might have snuck in from the previous PR
```suggestion
			$source_data['type'] ?? '',
```

File: plugins/woocommerce/src/Internal/Orders/OrderAttributionController.php
Comment (layoutd) [inline plugins/woocommerce/src/Internal/Orders/OrderAttributionController.php]: The ternary isn't necessary here
```suggestion
				'allowTracking' => 'yes' === $allow_tracking,
```

File: plugins/woocommerce/src/Internal/Orders/OrderAttributionController.php
Comment (layoutd) [inline plugins/woocommerce/src/Internal/Orders/OrderAttributionController.php]: I think this might have snuck in from the previous PR
```suggestion
			$order->add_meta_data( $this->get_meta_prefixed_field( 'type' ), 'admin' );
```

**Review by layoutd — CHANGES_REQUESTED** [review]
Thanks @tomalec! It looks good and works well in testing with shortcode and block checkout. I made a few suggestions, and few other comments:
- 🚧  We should add some code to prevent storing a [bunch of empty values](https://github.com/woocommerce/woocommerce/assets/228780/b4b9fe8f-defc-4ad5-94b8-8897bd530e28) in the meta table when consent isn't present. At the beginning of `OrderAttributionController::set_order_source_data()`:
```
	// If all the values are empty, bail.
	if ( empty( array_filter( $source_data ) ) ) {
		return;
	}
```
That way, when `document.querySelectorAll('input[name^=wc_order_attribution_]').forEach((el) => console.log(el.value));` gives all empty values, they're just and we avoid order edit metabox funkiness:
image
- ❓ I wonder if we should do the same for tracks? Or do we like this kind of event data:
```
Array
(
    [order_id] => 392
    [type] => 
    [medium] => 
    [source] => 
    [device_type] => (unknown)
    [origin_label] => unknown
    [session_pages] => 
    [session_count] => 
    [order_total] => 48.00
    [customer_order_count] => 1
    [customer_registered] => no
)
```
For now I'd continue sending tracks even if they're empty, and later on we can decide to filter them for stats or no send the event if the fields are empty.
- ❓ I'm not really sure if the "give consent after the first page" scenario is that important - by that point, the referer has been lost so the data is always Direct/typein. What's the actual rule here, until consent is given I guess we can't keep any information at all? Or is it just that we can't store it? I assume it's the latter since there must be some way to store the items in the cart from page to page, even if consent is denied. Could we keep the values in sourcebuster, but just not store/send them if consent is denied? Or does that blur the line between "functional" and "statistics" permissions?
- ✔️  For some reason my complianz shows the consent request on every page, so I was able to test the "Deny on checkout" scenario and it worked.

File: plugins/woocommerce/src/Internal/Orders/OrderAttributionController.php
Comment (layoutd) [reply plugins/woocommerce/src/Internal/Orders/OrderAttributionController.php]: OK, now I get the change, but it's not necessary since `OrderAttributionMeta::get_meta_prefixed_field()` specifically changes `type` to `source_type` before prefixing:
https://github.com/woocommerce/woocommerce/blob/cb0c42180a5860e6dd3c78b0ff004b5e9b20d1b8/plugins/woocommerce/src/Internal/Traits/OrderAttributionMeta.php#L161-L164
So passing `source_type` or `source` are essentially the same.

File: plugins/woocommerce/src/Internal/Orders/OrderAttributionController.php
Comment (layoutd) [reply plugins/woocommerce/src/Internal/Orders/OrderAttributionController.php]: This one will actually affect the Tracks reporting since `$source_data` has `type` but no `source_type`. So the `origin` label always ends up being `unknown`

File: plugins/woocommerce/src/Internal/Orders/OrderAttributionController.php
Comment (layoutd) [reply plugins/woocommerce/src/Internal/Orders/OrderAttributionController.php]: Reverted in 4992aaeb65d3b2552bdd7cf4f3979aa429625eeb

File: plugins/woocommerce/src/Internal/Orders/OrderAttributionController.php
Comment (layoutd) [reply plugins/woocommerce/src/Internal/Orders/OrderAttributionController.php]: Reverted in 4992aaeb65d3b2552bdd7cf4f3979aa429625eeb

File: plugins/woocommerce/src/Internal/Orders/OrderAttributionController.php
Comment (layoutd) [reply plugins/woocommerce/src/Internal/Orders/OrderAttributionController.php]: Removed in a289aa7013602cde94c88ffe4e1cd060edeca954

**Comment by layoutd** [comment]
> * We should add some code to prevent storing a [bunch of empty values](https://github.com/woocommerce/woocommerce/assets/228780/b4b9fe8f-defc-4ad5-94b8-8897bd530e28) in the meta table when consent isn't present. At the beginning of `OrderAttributionController::set_order_source_data()`:

Done in 00b83eed003cce1f69fdd7a7947c8908a0ad221a. Tracks data still send empty values.

If this passes tests before code freeze (seems unlikely) then I'll merge it since it makes a few more fixes and will only affect sites with a WP Consent API-compliant plugin.

**Review by layoutd — APPROVED** [review]
Tested well with block checkout, classic checkout, and shortcode checkout.

> * We should add some code to prevent storing a [bunch of empty values](https://github.com/woocommerce/woocommerce/assets/228780/b4b9fe8f-defc-4ad5-94b8-8897bd530e28) in the meta table when consent isn't present. At the beginning of `OrderAttributionController::set_order_source_data()`:

Done in 00b83eed003cce1f69fdd7a7947c8908a0ad221a. Tracks data still send empty values.

If this passes tests before code freeze (seems unlikely) then I'll merge it since it makes a few more fixes and will only affect sites with a WP Consent API-compliant plugin.

**Comment by tomalec** [comment]
Thanks for the quick review and additional fixes @layoutd!

Seems we didn't make it before the code freeze :(

> We should add some code to prevent storing a [bunch of empty values](https://github.com/woocommerce/woocommerce/assets/228780/b4b9fe8f-defc-4ad5-94b8-8897bd530e28) in the meta table when consent isn't present. At the beginning of OrderAttributionController::set_order_source_data():

Personally, I've seen some value in sending empty props instead of not sending anything.
- not sending a property means the data is not expected to be collected - for example, the hook was used to stop collecting a given property
- sending empty means this particular property was expected to be collected, but a user didn't give consent.

>  wonder if we should do the same for tracks?

Well, I think the above rationale stands for tracks as well. The difference is that track records are stored in our DBs and do not occupy merchant's DB with empty fields.

**Comment by layoutd** [comment]
> Well, I think the above rationale stands for tracks as well. The difference is that track records are stored in our DBs and do not occupy merchant's DB with empty fields.

Yes, I can see the value in seeing that consent was not given in the Tracks event, although I think it would probably be more elegant to have a specific flag for that scenario, instead of/accompanying those empty values. For the local store database, I guess something similar would also work.

**Comment by tomalec** [comment]
I'm marking it as [high-impact](https://github.com/woocommerce/woocommerce/blob/trunk/docs/contributing/deciding-pr-high-impact.md#my-pr-is-high-impact-whats-next) as:
- It's a big PR (i.e. adds several changes in many files).
- Modifies critical functionality: Shopper | Checkout | Existing customer can place order

File: plugins/woocommerce/src/Internal/Integrations/WPConsentAPI.php
Comment (tomalec) [inline plugins/woocommerce/src/Internal/Integrations/WPConsentAPI.php]: Preferably, I'd add a single data attribute to the `<script>` enqueued above, like:

```js
// plugins/woocommerce/src/Internal/Integrations/WPConsentAPI.php:90
wp_enqueue_script(
			'wp-consent-api-integration',
			plugins_url(
				"assets/js/frontend/wp-consent-api-integration{$this->get_script_suffix()}.js",
				WC_PLUGIN_FILE
			),
			array( 'wp-consent-api', 'wc-order-attribution' ),
			Constants::get_constant( 'WC_VERSION' ),
			true,
			array( 'data' => array( 'category' => 'marketing' ) )
		);
```
to produce 
```html
<script src="..." id="wp-consent-api-integration-js" data-category="marketing"></script>
```
which then would be used as
```js
// plugins/woocommerce/client/legacy/js/frontend/wp-consent-api-integration.js:8
document.currentString.dataset.category
```


But for some reason WP API limits that and forces us to use globals.

**Comment by alopezari** [comment]
Hi there! I've seen this PR has been approved for a while. Just wanted to share that you no longer need to wait for the release team to merge this PR even if it's high impact and you label it with `needs: analysis`. This also applies to [this one](https://github.com/woocommerce/woocommerce/pull/41856) and [this one](https://github.com/woocommerce/woocommerce/pull/41690).

However, thank you for flagging this for analysis! I agree they are indeed high impact, and the testing instructions look more complex than other PRs, so we will test them carefully.

You can merge them whenever you feel ready. Thanks!

**Pull request merged**
